ethdb/pebble/pebble.go

```diff
@@ -0,0 +1,571 @@
+// key-value data store.
+func (snap *snapshot) Get(key []byte) ([]byte, error) {
+	dat, closer, err := snap.db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]byte, len(dat))
+	copy(ret, dat)
+	closer.Close()
+	return ret, nil
+}
+
+// Release releases associated resources. Release should always succeed and can
+// be called multiple times without causing error.
+func (snap *snapshot) Release() {
+	snap.db.Close()
+}
+
+// upperBound returns the upper bound for the given prefix
+func upperBound(prefix []byte) (limit []byte) {
+	for i := len(prefix) - 1; i >= 0; i-- {
+		c := prefix[i]
+		if c == 0xff {
+			continue
+		}
+		limit = make([]byte, i+1)
+		copy(limit, prefix)
+		limit[i] = c + 1
+		break
+	}
+	return limit
+}
+
+// NewIterator creates a binary-alphabetical iterator over a subset
+// of database content with a particular key prefix, starting at a particular
+// initial key (or after, if it does not exist).
+func (d *Database) NewIterator(prefix []byte, start []byte) ethdb.Iterator {
+	iter := d.db.NewIter(&pebble.IterOptions{
+		LowerBound: append(prefix, start...),
+		UpperBound: upperBound(prefix),
+	})
+	iter.First()
+	return &pebbleIterator{iter: iter, moved: true}
+}
+
+// Stat returns a particular internal stat of the database.
+func (d *Database) Stat(property string) (string, error) {
+	return "", nil
+}
+
+// Compact flattens the underlying data store for the given key range. In essence,
+// deleted and overwritten versions are discarded, and the data is rearranged to
+// reduce the cost of operations needed to access them.
+//
+// A nil start is treated as a key before all keys in the data store; a nil limit
+// is treated as a key after all keys in the data store. If both is nil then it
+// will compact entire data store.
+func (d *Database) Compact(start []byte, limit []byte) error {
+	return d.db.Compact(start, limit, true) // Parallelization is preferred
+}
+
+// Path returns the path to the database directory.
+func (d *Database) Path() string {
+	return d.fn
+}
+
+// meter periodically retrieves internal pebble counters and reports them to
+// the metrics subsystem.
+func (d *Database) meter(refresh time.Duration) {
+	var errc chan error
+	timer := time.NewTimer(refresh)
+	defer timer.Stop()
+
+	// Create storage and warning log tracer for write delay.
+	var (
+		compTimes        [2]int64
+		writeDelayTimes  [2]int64
+		writeDelayCounts [2]int64
+		compWrites       [2]int64
+		compReads        [2]int64
+
+		nWrites [2]int64
+	)
+
+	// Iterate ad infinitum and collect the stats
+	for i := 1; errc == nil; i++ {
+		var (
+			compWrite int64
+			compRead  int64
+			nWrite    int64
+		)
+
+		metrics := d.db.Metrics()
+		compTime := atomic.LoadInt64(&d.compTime)
+		writeDelayCount := atomic.LoadInt64(&d.writeDelayCount)
+		writeDelayTime := atomic.LoadInt64(&d.writeDelayTime)
+		nonLevel0CompCount := int64(atomic.LoadUint32(&d.nonLevel0Comp))
+		level0CompCount := int64(atomic.LoadUint32(&d.level0Comp))
```

Comment:
Nitpick, if you place these in a `var (...)` group,they will be nicely aligned, like above